### PR TITLE
:sparkles: add job feature

### PR DESCRIPTION
**Pull request by Sagensagen**
# Job announcements
:sparkles: Adds all endpoints needed to CRUD jobs
:seedling: Added  seeding(without images)
:construction: Added general unit tests for jobs

### CODE REVIEW

File: app/api/jobs.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+@router.get('/')
+def get_jobs(request: Request):
+    db = get_database(request)
+    jobs = db.jobs.find()
+    return [JobItem.parse_obj(job) for job in jobs]
+
+
+@router.get('/{id}')
+def get_jobs(request: Request, id: str):
```
Comment (FredrikMorstad) [inline app/api/jobs.py:26]: get_job_by_id is more descriptive imo

File: db/seeds/jobs.json [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
... applikasjonssikkerhetskompetanse? Vi søker flere utviklere som kan være med i våre team, og da med et særskilt fokus på sikkerheten i livssyklusen (S-SDLC).  Hvorfor Bouvet ? 🥕 Vi har en sterk delingskultur. Det gjennomsyrer måten vi jobber sammen i prosjektene på, hvordan vi jobber mot et felles mål, våre lite spisse albuer, og at vi hjelper og deler kunnskap med hverandre. 🥕 Vi har byttet ut karrierestigen med en faglig rulletrapp. Ta steget,og vi vil gi deg kurset, mentoren eller sertifiseringen du trenger for å lære noe nytt. Hos oss er kunnskap og erfaring veien til nye prosjekter og mer ansvar. 🥕 Alle ansatte har fast lønn knyttet til stillingen sin. Prestasjonsbasert lønn og individuelle bonuser er ikke noe for oss. Vi vil gi deg trygghet med fast lønn og overskuddsdeling (felles bonusordning), og du får selvsagt betalt for overtid i perioder du må stå på litt ekstra. 🥕 Bærekraft er noe vi tar på alvor. Vår visjon «Vi går foran og bygger fremtidens samfunn» viser at vi ønsker å bidra til langsiktig verdiskapning med fokus på miljø og klima, sosiale og økonomiske forhold. Våre medarbeidere driver oss videre med nødvendig utålmodighet og krever at vi holder orden i eget hus",
```
Comment (FredrikMorstad) [inline db/seeds/jobs.json:46]: Jeg ser det er brukt emojis, hvordan går det når det parseres i frontend?

File: pytest_docker.py [hunk 3]
```diff
@@ -3,13 +3,15 @@
 import sys
```
Comment (FredrikMorstad) [inline pytest_docker.py:3]: I think these changes should be in separate commit since they are just formatting

File: tests/conftest.py [hunk 4]
```diff
@@ -7,9 +7,11 @@
```
Comment (FredrikMorstad) [inline tests/conftest.py:7]: same here add it in a separate commit as a format commit

File: tests/test_endpoints/test_jobs.py [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+@admin_required("/api/jobs/", "post")
+def test_create_job(client):
+    # Test admin auth
+    access_token = client_login(
+        client, regular_member["email"], regular_member["password"])
+    headers = {"Authorization": f"Bearer {access_token}"}
+    response = client.post(
+        "/api/jobs/", json=new_job, headers=headers)
+    assert response.status_code == 403
```
Comment (FredrikMorstad) [inline tests/test_endpoints/test_jobs.py:36]: if thees just test admin credential, the admin required decorator does exactly that

File: app/api/jobs.py [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+
+@router.post('/')
+def create_job(request: Request, job: JobItemPayload, token: AccessTokenPayload = Depends(authorize_admin)):
+
+    db = get_database(request)
+    item = job.dict()
+    item['id'] = uuid4()
+    item['published_date'] = datetime.now()
+    _job = JobItem.parse_obj(item)
+    db.jobs.insert_one(_job.dict())
```
Comment (FredrikMorstad) [inline app/api/jobs.py:42]: maybe a 500 if insert did not work

File: app/api/jobs.py [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+
+
+@router.put('/{id}')
+def update_job(request: Request, id: str, job: JobItemPayload, token: AccessTokenPayload = Depends(authorize_admin)):
+    db = get_database(request)
+    original_job = db.jobs.find_one({'id': UUID(id)})
+    if not original_job:
+        raise HTTPException(400, "Job could not be found")
+    db.jobs.update_one({'id': UUID(id)}, {
+        '$set': job.dict()})
```
Comment (FredrikMorstad) [inline app/api/jobs.py:63]: How does this handle empty strings or undefined for required fields? I think the way its done in update event is more suitable to handle bad requests

File: tests/test_endpoints/test_jobs.py [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+
+    access_token = client_login(
+        client, admin_member["email"], admin_member["password"])
+
+    headers = {"Authorization": f"Bearer {access_token}"}
+    response = client.post(
+        "/api/jobs/", json=new_job, headers=headers)
+    assert response.status_code == 200
```
Comment (FredrikMorstad) [inline tests/test_endpoints/test_jobs.py:44]: should actually check if the db entry is inside the db after a 200

File: tests/test_endpoints/test_jobs.py [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+
+
+def test_get_job(client):
+    access_token = client_login(
+        client, admin_member["email"], admin_member["password"])
+    headers = {"Authorization": f"Bearer {access_token}"}
+    # insert new Job
```
Comment (FredrikMorstad) [inline tests/test_endpoints/test_jobs.py:51]: Can be an idea to seed the events in conftest we don't have to create a new event

File: tests/conftest.py [hunk 10]
```diff
@@ -35,8 +40,10 @@ def client(app):
         seed_events(app.db, f"{test_seed_path}/test_events.json")
```
Comment (FredrikMorstad) [inline tests/conftest.py:40]: add seed_jobs here? So all seeding events are available in every test

File: app/models.py [hunk 11]
```diff
@@ -166,14 +168,49 @@ class ChangePasswordPayload(BaseModel):
...
+    type: str
+    img: str
+    tags: List[str]
+    description_preview: str
+    description: str
+    start_date: datetime
+    published_date: datetime
+    location: str
+    link: str
+    due_date: datetime
```
Comment (FredrikMorstad) [inline app/models.py:216]: An idea here if here could be:
```
 Class JobItemPayload(BaseModel):
    # all fields
  Class JobItem(JobItemPayload):
    id: UUID4
```
Meaning JobItem always extend JobItemPayload

File: seeding.py [hunk 12]
```diff
@@ -114,8 +114,21 @@ def get_db():
     return MongoClient(conf.MONGO_URI, uuidRepresentation="standard")[conf.MONGO_DBNAME]
 
 
+def seed_jobs(db, seed_path):
+    with open(seed_path, "r") as f:
+        jobs = json.load(f)
+
+    for job in jobs:
```
Comment (FredrikMorstad) [inline seeding.py:121]: if we add an static uuid in the seeding we can check if it exists in the db already and avoid inserting the jobs multiple times

File: tests/test_endpoints/test_jobs.py [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+from tests.users import regular_member, admin_member
+
+from tests.utils.authentication import admin_required, authentication_required
+
+db = get_test_db()
+
+new_job = {
+    "company": "TD",
+    "title": "Vil du være TD's nye nettside ansvarlig?",
+    "type": "Charity",
```
Comment (FredrikMorstad) [inline tests/test_endpoints/test_jobs.py:15]: You'll get free coffee :laughing:

File: app/api/jobs.py [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+    return {'id': id}
+
+
+@router.put('/{id}')
+def update_job(request: Request, id: str, job: JobItemPayload, token: AccessTokenPayload = Depends(authorize_admin)):
+    db = get_database(request)
+    original_job = db.jobs.find_one({'id': UUID(id)})
+    if not original_job:
+        raise HTTPException(400, "Job could not be found")
+    db.jobs.update_one({'id': UUID(id)}, {
```
Comment (seblj) [inline app/api/jobs.py:62]: I think there is find_one_and_update so that we can avoid find_one and then do delete_one later

File: app/api/jobs.py [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+    return {'id': _job.id}
+
+
+@router.delete('/{id}')
+def delete_job(request: Request, id: str, token: AccessTokenPayload = Depends(authorize_admin)):
+    db = get_database(request)
+    job = db.jobs.find_one({'id': UUID(id)})
+    if not job:
+        raise HTTPException(400, "Job could not be found")
+    db.jobs.delete_one({'id': job['id']})
```
Comment (seblj) [inline app/api/jobs.py:52]: I think there also is find_one_and_delete. Is it possible to use this instead?

File: tests/test_endpoints/test_jobs.py [hunk 16]
```diff
@@ -0,0 +1,114 @@
...
+    response = response.json()
+
+    unauth_respoonse = client.delete(
+        "/api/jobs/"+response["id"], headers=headers_regular)
+
+    assert unauth_respoonse.status_code == 403
+
+    delResponse = client.delete(
+        "/api/jobs/"+response["id"], headers=headers_admin)
+    assert delResponse.status_code == 200
```
Comment (FredrikMorstad) [inline tests/test_endpoints/test_jobs.py:86]: same here should check if the job is actually deleted

File: tests/test_endpoints/test_jobs.py [hunk 17]
```diff
@@ -0,0 +1,114 @@
...
+    headers_regular = {"Authorization": f"Bearer {access_token_regular}"}
+
+    response = client.post(
+        "/api/jobs/", json=new_job, headers=headers_admin)
+    assert response.status_code == 200
+    response = response.json()
+
+    unauth_respoonse = client.delete(
+        "/api/jobs/"+response["id"], headers=headers_regular)
+
```
Comment (FredrikMorstad) [inline tests/test_endpoints/test_jobs.py:81]: same here use the admin required decorator

File: seeding.py [hunk 18]
```diff
@@ -114,8 +114,21 @@ def get_db():
...
+        job["id"] = uuid4()
+        job['start_date'] = datetime.now()
+        job['published_date'] = datetime.now()
+        job["due_date"] = datetime.now()
+        db.jobs.insert_one(job)
+
```
Comment (seblj) [inline seeding.py:127]: Probably not that important, but i think it would be a bit faster to create the objects and use `insert_many` or something

File: tests/test_endpoints/test_jobs.py [hunk 19]
```diff
@@ -0,0 +1,114 @@
...
+    response = client.post(
+        "/api/jobs/", json=new_job, headers=headers_admin)
+    assert response.status_code == 200
+    response = response.json()
+
+    new_job_update = new_job.copy()
+    new_job_update["title"] = "New Title"
+    update_response = client.put(
+        "/api/jobs/"+response["id"], json=new_job_update, headers=headers_regular)
+    assert update_response.status_code == 403
```
Comment (FredrikMorstad) [inline tests/test_endpoints/test_jobs.py:106]: same here use the decorator

**Review by FredrikMorstad — CHANGES_REQUESTED** [review]
The functionality looks good, however there are somethings i think you should look at :smile: Also a test for getting the job images would be nice

File: app/api/jobs.py (hunk 15 above)
Comment (Sagensagen) [reply app/api/jobs.py:52]: Will look into it :+1:

File: seeding.py (hunk 18 above)
Comment (Sagensagen) [reply seeding.py:127]: Absolutely. Did just this to optimize another mongodb app, and was way faster :zap:

File: db/seeds/jobs.json (hunk 2 above)
Comment (Sagensagen) [reply db/seeds/jobs.json:46]: Det går bra :) Er bare test seeds anyway

**Comment by Sagensagen** [comment]
Should it be implicit that each JobItem should have an image? If not found it will simply not display any

**Comment by Sagensagen** [comment]
Missed out on a couple comments. Stand  by :D

**Review by seblj — APPROVED** [review]
Looks good for the things I commented on🙌

File: app/models.py (hunk 11 above)
Comment (Sagensagen) [reply app/models.py:216]: Hahah don't know why i did this :smile:

File: tests/test_endpoints/test_jobs.py [hunk 20]
```diff
@@ -0,0 +1,109 @@
...
+    new_job_update = new_job.copy()
+    new_job_update["title"] = "New Title"
+
+    update_response = client.put(
+        "/api/jobs/"+response["id"], json=new_job_update, headers=headers)
+    assert update_response.status_code == 200
+
+    _test = client.get("/api/jobs/"+response["id"], headers=headers)
+
+    assert _test.json()["title"] == new_job_update["title"]
```
Comment (FredrikMorstad) [inline tests/test_endpoints/test_jobs.py:109]: nice nice

**Review by FredrikMorstad — APPROVED** [review]
